src/atomic/atomic_cell.rs

```diff
@@ -0,0 +1,695 @@
+    /// a.set(8);
+    /// assert_eq!(a.get(), 8);
+    /// ```
+    pub fn set(&self, val: T) {
+        if mem::needs_drop::<T>() {
+            drop(self.replace(val));
+        } else {
+            unsafe {
+                atomic_store(self.value.get(), val);
+            }
+        }
+    }
+
+    /// Stores `val` into the atomic cell and returns the previous value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// let a = AtomicCell::new(7);
+    ///
+    /// assert_eq!(a.get(), 7);
+    /// assert_eq!(a.replace(8), 7);
+    /// assert_eq!(a.get(), 8);
+    /// ```
+    pub fn replace(&self, val: T) -> T {
+        unsafe { atomic_swap(self.value.get(), val) }
+    }
+}
+
+impl<T: Default> AtomicCell<T> {
+    /// Takes the inner value and replaces it with `T::default()`.
+    ///
+    /// Note that `atomic_cell.take()` is equivalent to:
+    ///
+    /// ```ignore
+    /// atomic_cell.replace(T::default())
+    /// ```
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// let a = AtomicCell::new(7);
+    ///
+    /// assert_eq!(a.get(), 7);
+    /// assert_eq!(a.take(), 7);
+    /// assert_eq!(a.get(), 0);
+    /// ```
+    pub fn take(&self) -> T {
+        self.replace(T::default())
+    }
+}
+
+impl<T: Copy> AtomicCell<T> {
+    /// Returns a copy of the inner value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// let a = AtomicCell::new(7);
+    ///
+    /// assert_eq!(a.get(), 7);
+    /// ```
+    pub fn get(&self) -> T {
+        unsafe { atomic_load(self.value.get()) }
+    }
+
+    /// Updates the inner value using a function and returns the new value.
+    ///
+    /// Function `f` might have to be called multiple times if the inner value is concurrently
+    /// changed by other threads.
+    ///
+    /// Note that `atomic_cell.update(f)` is equivalent to:
+    ///
+    /// ```ignore
+    /// loop {
+    ///     let current = atomic_cell.get();
+    ///     let new = f(current);
+    ///
+    ///     if atomic_cell.compare_and_set(current, new) {
+    ///         break new;
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// let a = AtomicCell::new(7);
+    ///
+    /// assert_eq!(a.update(|x| x.min(9)), 7);
+    /// assert_eq!(a.update(|x| x.min(5)), 5);
+    ///
+    /// a.update(|x| x * 10);
+    /// assert_eq!(a.get(), 50);
+    /// ```
+    pub fn update<F>(&self, mut f: F) -> T
+    where
+        F: FnMut(T) -> T,
+    {
+        let mut current = self.get();
+
+        loop {
+            let new = f(current);
+
+            let previous = unsafe {
+                atomic_compare_and_swap(self.value.get(), current, new)
+            };
+
+            if byte_eq(&previous, &current) {
+                return new;
```

Comment:
The general convention for atomic operations is to return the previous value.

Reply:
However, if you change the return value of `F` to `Option<T>` then it would make more sense to return a `bool` here.